### ISSUE TITLE
Intermittent false negatives (silently exiting without any output)

### ISSUE THREAD

**Issue by karlhorky** [remarkjs/remark-lint-no-dead-urls#53]
### Initial checklist

- [X] I read the [support docs](https://github.com/remarkjs/.github/blob/main/support.md)
- [X] I read the [contributing guide](https://github.com/remarkjs/.github/blob/main/contributing.md)
- [X] I agree to follow the [code of conduct](https://github.com/remarkjs/.github/blob/main/code-of-conduct.md)
- [X] I searched [issues](https://github.com/search?q=user%3Aremarkjs&type=Issues) and couldn’t find anything (or linked relevant results below)

### Affected packages and versions

remark-cli@12.0.1, remark-lint@10.0.0, remark-lint-no-dead-urls@2.0.0

### Link to runnable example

https://github.com/karlhorky/repro-remark-lint-no-dead-urls-intermittent-silent-failures

### Steps to reproduce

1. Run the linting script multiple times in a row with the `.mdx` content in https://github.com/karlhorky/repro-remark-lint-no-dead-urls-intermittent-silent-failures/blob/main/src/pages/a/c/y.mdx
2. 💥 Observe that the run results are different, with some of the runs exiting with code 0 silently without any output, indicating some kind of silent crash, see [test run](https://github.com/karlhorky/repro-remark-lint-no-dead-urls-intermittent-silent-failures/actions/runs/11092568056/job/30817779214) (the 1st run exits without any output, the 2nd run fails as it should):
   ![Screenshot 2024-09-29 at 14 26 12](https://github.com/user-attachments/assets/4123f358-eb05-4ab6-97bd-53a71eb88163)

Originally investigated in https://github.com/remarkjs/remark-lint-no-dead-urls/issues/52

Runtime: Latest Node.js LTS (v20.17.0)

### Expected behavior

`remark-lint-no-dead-urls` checks all URLs in all specified files and shows the results

### Actual behavior

`remark-lint-no-dead-urls` exits with code 0 and no output in some runs

### Runtime

Other (please specify in steps to reproduce)

### Package manager

pnpm

### OS

Linux

**Comment by wooorm** [comment]
Thanks, I can reproduce!

* For the types:

  ```diff
  -/** @type {Omit<import('unified-engine').Options, 'processor'>} */
  +/** @type {Omit<import('unified-engine').Preset} */
  ```
* I would always recommend making your reproductions as small as possible: there’s a lot of `skipUrlPatterns`, but I can reproduce the problem without them. And the code block isn’t it either. Or `x.mdx`, or the folder structure, or `remark-mdx`.
* It does seem to the many URLs. Less URLs seems to make it appear less often

**Comment by wooorm** [comment]
I checked which URLs failed, out of the 50. The 4 ones that never came back were to `https://www.youtube.com/`. There were in total 6 connections to `https://www.youtube.com/`. That seems to indicate that youtube does some DDoS prevention

**Comment by wooorm** [comment]
It seems to be that `response.text()` just hangs:

https://github.com/wooorm/dead-or-alive/blob/42d293a850870b507862c6ae318501a6fd2858be/lib/index.js#L329

Seems to be related to the signal. No hang happens without an abort signal:

https://github.com/wooorm/dead-or-alive/blob/42d293a850870b507862c6ae318501a6fd2858be/lib/index.js#L210

**Comment by wooorm** [comment]
I’m thinking there’s some race condition in play as well. Perhaps YouTube is, when it detects many requests, slow in sending the HTML body. The request went fine but the body is slow and then aborted. This may be an undici bug though, that it just aborts silently, instead of throwing. No clue.

**Comment by wooorm** [comment]
released in [`dead-or-alive@1.0.3`](https://github.com/wooorm/dead-or-alive/releases/tag/1.0.3); `npm update` should do the trick.

**Comment by karlhorky** [comment]
Hmm, tried `dead-or-alive@1.0.3` now in my reproduction repo in a PR, and it still silently exited in the first step (the rest of the steps reported each a consistent 15 warnings):

- https://github.com/karlhorky/repro-remark-lint-no-dead-urls-intermittent-silent-failures/pull/1
- https://github.com/karlhorky/repro-remark-lint-no-dead-urls-intermittent-silent-failures/actions/runs/11110440262/job/30868135467?pr=1

This PR also adds `|| true` to enable running the same command multiple times without the workflow run failing.

Maybe still intermittently failing for another reason?

I can reduce the reproduction to simplify if it would help.

![Screenshot 2024-09-30 at 18 41 09](https://github.com/user-attachments/assets/520c2ddc-95d3-47ff-881b-74dc495d6db9)

**Comment by ChristianMurphy** [comment]
> I can reduce the reproduction to simplify if it would help.

Please do, thanks!

**Comment by karlhorky** [comment]
I simplified the repro now:

- https://github.com/karlhorky/repro-remark-lint-no-dead-urls-intermittent-silent-failures/pull/2

![Screenshot 2024-10-01 at 10 59 45](https://github.com/user-attachments/assets/50392587-d493-4d76-b7ed-7bd0328059a2)

Failed the first time, but then after merging to the `dead-or-alive@1.0.3` branch, it's harder to reproduce (a few test runs now without the silently exiting behavior).

It's not `dead-or-alive@1.0.3` that fixed this though, because [I reproduced the behavior with `dead-or-alive@1.0.3` above](https://github.com/remarkjs/remark-lint-no-dead-urls/issues/53#issuecomment-2383689294) with the more complex repro. 

So maybe:

1. The extra complexity in the reproduction also caused another, unrelated situation that causes `remark-lint-no-dead-urls` to fail
2. It's a behavior of another of the services (eg. maybe `archive.is` - I see a 429 response there in the screenshot above) that only happens the first time after some period of inactivity.

I'll try re-running after a few minutes and few hours. And also try adding back parts of the original, more complex reproduction, in case those were indeed relevant.

**Comment by karlhorky** [comment]
Yeah, it seems like the 2nd file `x.mdx` was indeed a relevant part of it (or at least, it causes the problem to appear right away):

- Commit: [karlhorky/repro-remark-lint-no-dead-urls-intermittent-silent-failures@`9d0bd55` (#1)](https://github.com/karlhorky/repro-remark-lint-no-dead-urls-intermittent-silent-failures/pull/1/commits/9d0bd555570d4e181ba8b871ea500ea6d7298261)
- Workflow run: https://github.com/karlhorky/repro-remark-lint-no-dead-urls-intermittent-silent-failures/actions/runs/11122491557/job/30903865249?pr=1

Interestingly, it is now very consistent in its failures:

![Screenshot 2024-10-01 at 11 27 24](https://github.com/user-attachments/assets/9242d378-4926-4888-a79f-b70b0cab6091)

**Comment by karlhorky** [comment]
Hm... one other change that I did while copying the original reproduction back, was this change:

```diff
/** @type {import('unified-engine').Preset} */
const config = {
-  plugins: ['remark-mdx', 'remark-lint', 'remark-lint-no-dead-urls'],
+  plugins: [
+    'remark-mdx',
+    [
+      'remark-lint-no-dead-urls',
+      {
+        skipUrlPatterns: [
+          'chrome://',
+          'codesandbox-link://',
+          'embedded-html-codesandbox://',
+          'embedded-css-codesandbox://',
```

Maybe this omission of `'remark-lint'` is the problem? I thought I tested `dead-or-alive@1.0.3` with `'remark-lint'` in the `plugins`, but maybe not...

**Edit:** No, I guess not, still intermittently exiting silently with `'remark-lint'` in plugins:

- Commit: [karlhorky/repro-remark-lint-no-dead-urls-intermittent-silent-failures@`e025ef9` (#1)](https://github.com/karlhorky/repro-remark-lint-no-dead-urls-intermittent-silent-failures/pull/1/commits/e025ef934f4b1a62bb43d7ad23c24069c24cd34c)
- Workflow run: https://github.com/karlhorky/repro-remark-lint-no-dead-urls-intermittent-silent-failures/actions/runs/11122635695/job/30904325260?pr=1

![Screenshot 2024-10-01 at 11 37 06](https://github.com/user-attachments/assets/246d6734-5421-443f-b1e4-f1cf43ef7474)

**Comment by karlhorky** [comment]
No, I guess not, still intermittently exiting silently with `'remark-lint'` in plugins:

- Commit: [karlhorky/repro-remark-lint-no-dead-urls-intermittent-silent-failures@`e025ef9` (#1)](https://github.com/karlhorky/repro-remark-lint-no-dead-urls-intermittent-silent-failures/pull/1/commits/e025ef934f4b1a62bb43d7ad23c24069c24cd34c)
-

**Comment by wooorm** [comment]
I did not reproduce anything else. If you want me to spend time on this, can you please please make a tiny reproduction that I can work with? Removing everything unrelated?

**Comment by karlhorky** [comment]
[The `dead-or-alive-1.0.3` branch](https://github.com/karlhorky/repro-remark-lint-no-dead-urls-intermittent-silent-failures/tree/dead-or-alive-1.0.3/src/pages) ([PR 1 in my reproduction repo](https://github.com/karlhorky/repro-remark-lint-no-dead-urls-intermittent-silent-failures/pull/1)) is the smallest reproduction which I have been able to consistently reproduce the problem with.

To me, this seems pretty minimal - but maybe you have another opinion on what should still be simplified here.

It's:

1. 1 remark config file
2. 1 npm script for anything I cannot specify in the remark config
3. 2 MDX files (as mentioned above - 2 files are necessary, it seems)
4. CI setup to run the npm script

![Screenshot 2024-10-01 at 12 10 34](https://github.com/user-attachments/assets/cf9172c6-78b0-40a3-bae5-3e5c6ca7050a)

**Comment by karlhorky** [comment]
@wooorm I'm really not sure what I should change to make this more minimal, but happy to learn if you have some ideas.

Should this issue be reopened, since it's still failing in the same way? Or I could open this as a new issue if that's preferred.

**Comment by wooorm** [comment]
Here’s how I work through your code:

- custom package manager is not needed, builtin npm repros too
- TS is not used, drop unified-engine
- mdx is not needed for your repro, rename `a/b/y.mdx` -> `a/b/y.md`, `c/x.mdx` -> `c/x.md`, remove `remark-mdx` dependency, remove `--ext mdx` flag, remove `remark-mdx` in your config file
- config file is not needed, replace `--rc-path remark-lint-no-dead-urls.mjs` with `--use remark-lint --use remark-lint-no-dead-urls`; remove `remark-lint-no-dead-urls.mjs`
- other flags aren’t needed: remove `--frail --no-stdout`; specific path is not needed: `./src/pages/` -> `.`
- remark-lint itself isn’t needed to reproduce either: drop `--use remark-lint`, drop the dependency
- complex file structure isn’t needed, `src/pages/a/b/y.md` -> `y.md`, `src/pages/c/x.md` -> `x.md`, remove `src/`
- removing the first 3 links in `x.md` is fine; removing all is not fine; removing the first 4 is fine.
  So, leaving it at solely:
  ```
  - [See How Fast ARPANET Spread in Just Eight Years](https://web.archive.org/web/20210506160300/https://www.smithsonianmag.com/smart-news/see-how-fast-arpanet-spread-in-just-eight-years-2341268/)
  ```
  (note the `archive.org`, perhaps relevant?)
- removing all the URLs from `y.md` that aren‘t `archive.org`/`archive.ph`: it works, so that’s not it.
- removing the first 10 instead? still repros. Next 10? Still repros. Next 10? Works again. Revert. 5? Repros. 2 more? Repros. 1 more? Repros. 1 more? Works again, revert.
  Here it gets vague and flakey. At about 10 URLs total; split over 2 files, repros.
- alright, so we narrowed it down a lot; we know that many connections at once repros things. Now on to the main hunch for a solution: https://github.com/remarkjs/remark-lint-no-dead-urls/issues/52#issuecomment-2380855457; make the connections concurrent; adding `p-all` with `{concurrency: 3}` to in `node_modules` and suddenly it’s solved; reverting those markdown files so they have their 5 and  like 55 URLs again, that works too. So, that’s it. `concurrency: 3` and `concurrency: 4` are fine. `concurrency: 5` is not fine! We know the problem earlier was about 10 URLs over 2 files. `concurrency: 5` for 2 files is oddly familiar. So, seems we need to pool total connections. See above commit

**Comment by karlhorky** [comment]
Wow nice, that's a great simplification, thanks for the deep dive!

I think I'm not versed enough on how to use `remark-cli` flags, and was not aware that `remark-lint` could be dropped, I thought it was a dependency of `remark-lint-no-dead-urls` 👀

Next time that I do a repro in any of the remark / unified things, I'll be sure to try out simpler combinations of packages using these steps.

**Comment by karlhorky** [comment]
I can confirm that the silent exiting behavior is no longer reproducible in my first 5 test runs, after I upgraded to `remark-lint-no-dead-urls@2.0.1` in my reproduction repo:

- https://github.com/karlhorky/repro-remark-lint-no-dead-urls-intermittent-silent-failures/pull/3

![Screenshot 2024-10-08 at 18 06 38](https://github.com/user-attachments/assets/e3e1460e-2100-47f1-b13c-3e2a9e7b90b4)

**Comment by karlhorky** [comment]
Upgrading to `remark-lint-no-dead-urls@2.0.1` in the monorepo with 147 MDX files, I noticed timeouts with v2:

- `remark-lint-no-dead-urls@1.1.0` test runs for all 147 MDX files
  - ~1-4 minutes (multiple test runs over weeks)
- `remark-lint-no-dead-urls@2.0.1` test runs for all 147 MDX files
  - timed out at 26, 48, 48, 48, 48 minutes (first 5 test runs just now, timeout caused by GitHub Actions workflow `timeout-minutes` config)

I'll see if this continues, and if it does, I'll try to find time to create a minimal repro.
